## Hackathon Battle Data - part 1: EDA

1- First importing pandas and uploading the csv file

In [11]:
import pandas as pd 
df = pd.read_csv("segment_alerts_all_airports_train.csv", sep =",")

### Data Inspection

In [12]:
df.head()

,lightning_id,lightning_airport_id,date,lon,lat,amplitude,maxis,icloud,dist,azimuth,airport,airport_alert_id,is_last_lightning_cloud_ground
0,1,1,2016-01-02 14:53:36+00:00,9.0559,42.0826,-9.90,0.3,False,27.360653,57.852343,Ajaccio,NaN,NaN
1,2,2,2016-01-02 14:53:36+00:00,9.0236,42.0953,-3.33,0.2,True,26.383167,52.117828,Ajaccio,NaN,NaN
2,3,3,2016-01-02 21:22:53+00:00,8.8585,42.0456,-18.68,0.4,True,14.313391,24.500543,Ajaccio,NaN,NaN
3,4,4,2016-01-02 21:22:53+00:00,8.8517,42.0517,-7.51,0.2,False,14.794117,20.854458,Ajaccio,1.0,False
4,5,5,2016-01-02 21:24:46+00:00,8.8728,42.0494,-6.01,0.2,False,15.124224,29.058471,Ajaccio,1.0,False
5,6,6,2016-01-02 21:25:59+00:00,8.8495,42.0595,-23.96,2.6,False,15.583922,18.926800,Ajaccio,1.0,False
6,7,7,2016-01-02 21:27:04+00:00,8.8978,42.0423,-11.65,0.2,False,15.343433,38.642144,Ajaccio,1.0,False
7,8,8,2016-01-02 21:28:54+00:00,8.9269,42.0576,-10.97,0.8,True,18.071989,42.780344,Ajaccio,NaN,NaN
8,9,9,2016-01-02 21:28:54+00:00,8.9119,42.0593,-5.31,0.6,False,17.561774,38.772909,Ajaccio,1.0,True
9,10,10,2016-01-02 21:32:23+00:00,8.9725,42.0501,-43.09,0.2,True,19.844864,53.281750,Ajaccio,NaN,NaN


In [13]:
df.tail()

,lightning_id,lightning_airport_id,date,lon,lat,amplitude,maxis,icloud,dist,azimuth,airport,airport_alert_id,is_last_lightning_cloud_ground
507066,628722,156714,2022-12-11 08:24:49+00:00,10.6409,43.8337,-46.83,0.085,False,24.786731,60.171014,Pise,NaN,NaN
507067,628723,156715,2022-12-11 09:23:21+00:00,10.6227,43.7234,14.00,0.088,True,18.243447,82.764678,Pise,NaN,NaN
507068,628724,156716,2022-12-11 09:29:39+00:00,10.6658,43.7448,23.47,0.171,False,22.129609,79.427030,Pise,NaN,NaN
507069,628725,156717,2022-12-11 09:33:50+00:00,10.7072,43.7657,16.69,0.170,True,25.964114,77.080081,Pise,NaN,NaN
507070,628726,156718,2022-12-11 09:40:10+00:00,10.7275,43.7709,12.00,0.096,True,27.691506,76.990098,Pise,NaN,NaN


In [17]:
df.sample(5)

,lightning_id,lightning_airport_id,date,lon,lat,amplitude,maxis,icloud,dist,azimuth,airport,airport_alert_id,is_last_lightning_cloud_ground
53041,53042,53042,2020-09-23 08:14:43+00:00,8.6223,41.9894,-20.18,0.078,False,16.618701,290.018760,Ajaccio,355.0,False
292746,356369,94327,2021-09-02 18:35:09+00:00,-1.6116,43.3116,-8.81,0.063,True,18.794490,209.206500,Biarritz,NaN,NaN
415915,537571,65563,2018-07-06 08:25:02+00:00,10.3602,43.6064,-9.04,0.071,False,10.327657,203.649729,Pise,345.0,False
428118,549774,77766,2018-10-29 13:50:02+00:00,10.1633,43.6265,-4.89,3.177,True,20.419351,253.794885,Pise,NaN,NaN
147674,169006,75174,2019-07-15 11:18:17+00:00,9.4506,42.4967,3.49,0.155,True,6.787494,210.586120,Bastia,NaN,NaN


In [18]:
df.shape

(507071, 13)

In [22]:
list(df.columns)

['lightning_id',
 'lightning_airport_id',
 'date',
 'lon',
 'lat',
 'amplitude',
 'maxis',
 'icloud',
 'dist',
 'azimuth',
 'airport',
 'airport_alert_id',
 'is_last_lightning_cloud_ground']

In [23]:
df.dtypes

lightning_id                        int64
lightning_airport_id                int64
date                                  str
lon                               float64
lat                               float64
amplitude                         float64
maxis                             float64
icloud                               bool
dist                              float64
azimuth                           float64
airport                               str
airport_alert_id                  float64
is_last_lightning_cloud_ground     object
dtype: object

In [24]:
df.info

<bound method DataFrame.info of         lightning_id  lightning_airport_id                       date  \
0                  1                     1  2016-01-02 14:53:36+00:00   
1                  2                     2  2016-01-02 14:53:36+00:00   
2                  3                     3  2016-01-02 21:22:53+00:00   
3                  4                     4  2016-01-02 21:22:53+00:00   
4                  5                     5  2016-01-02 21:24:46+00:00   
...              ...                   ...                        ...   
507066        628722                156714  2022-12-11 08:24:49+00:00   
507067        628723                156715  2022-12-11 09:23:21+00:00   
507068        628724                156716  2022-12-11 09:29:39+00:00   
507069        628725                156717  2022-12-11 09:33:50+00:00   
507070        628726                156718  2022-12-11 09:40:10+00:00   

            lon      lat  amplitude  maxis  icloud       dist    azimuth  \
0        9.0559

In [34]:
## printing the variables of the dataset with the corresponding number of missing values found
df.isnull().sum()

lightning_id                           0
lightning_airport_id                   0
date                                   0
lon                                    0
lat                                    0
amplitude                              0
maxis                                  0
icloud                                 0
dist                                   0
azimuth                                0
airport                                0
airport_alert_id                  450472
is_last_lightning_cloud_ground    450472
dtype: int64

In [36]:
df.describe(include="all")

,lightning_id,lightning_airport_id,date,lon,lat,amplitude,maxis,icloud,dist,azimuth,airport,airport_alert_id,is_last_lightning_cloud_ground
count,507071.000000,507071.000000,507071,507071.000000,507071.000000,507071.000000,507071.000000,507071,507071.000000,507071.000000,507071,56599.000000,56599
unique,NaN,NaN,394839,NaN,NaN,NaN,NaN,2,NaN,NaN,5,NaN,2
top,NaN,NaN,2016-09-14 19:51:27+00:00,NaN,NaN,NaN,NaN,True,NaN,NaN,Pise,NaN,False
freq,NaN,NaN,24,NaN,NaN,NaN,NaN,378079,NaN,NaN,156718,NaN,53972
mean,319152.873651,59451.210188,NaN,6.385887,43.345451,-2.254868,1.025518,NaN,20.243757,168.806028,NaN,316.932720,NaN
std,192255.492060,39952.861744,NaN,5.196362,1.246482,20.252913,1.359688,NaN,6.941580,97.587294,NaN,189.621099,NaN
min,1.000000,1.000000,NaN,-2.006800,41.654300,-429.500000,0.016000,NaN,0.037212,0.000000,NaN,1.000000,NaN
25%,148099.500000,25354.000000,NaN,-1.295800,42.471800,-5.610000,0.107000,NaN,15.412370,85.240722,NaN,170.000000,NaN
50%,317158.000000,54199.000000,NaN,9.339100,43.445100,-1.560000,0.245000,NaN,21.442300,156.717299,NaN,282.000000,NaN
75%,501958.500000,90354.000000,NaN,10.220500,43.666200,4.620000,1.538000,NaN,26.112228,251.482192,NaN,446.000000,NaN


In [38]:
df["is_last_lightning_cloud_ground"].value_counts()

is_last_lightning_cloud_ground
False    53972
True      2627
Name: count, dtype: int64

In [40]:
df.nunique()

lightning_id                      507071
lightning_airport_id              156718
date                              394839
lon                                32954
lat                                26779
amplitude                          17080
maxis                               7587
icloud                                 2
dist                              504404
azimuth                           502666
airport                                5
airport_alert_id                     769
is_last_lightning_cloud_ground         2
dtype: int64

In [48]:
## Knowing the range of the continuous numerical variables
df.select_dtypes(include ='float').agg(['min','max'])

,lon,lat,amplitude,maxis,dist,azimuth,airport_alert_id
min,-2.0068,41.6543,-429.5,0.016,0.037212,0.000000,1.0
max,10.7717,47.4226,497.0,6.900,29.999996,359.978549,769.0


In [49]:
df["icloud"].value_counts(normalize=True)
## most of lightning events are intra-cloud lightning

icloud
True     0.745614
False    0.254386
Name: proportion, dtype: float64

In [50]:
df["airport"].value_counts()

airport
Pise        156718
Bastia      125919
Biarritz    115191
Ajaccio      72501
Nantes       36742
Name: count, dtype: int64

In [55]:
(df["amplitude"]>0).mean()

np.float64(0.4809858974384258)

In [58]:
## It is indicated that the last 2 columns are filled only if the dist <= 20km
df[df["airport_alert_id"].notnull()]["dist"].describe()

count    56599.000000
mean        13.513969
std          4.666377
min          0.037212
25%         10.310612
50%         14.373750
75%         17.413794
max         19.999984
Name: dist, dtype: float64

In [59]:
df.duplicated().sum()
## No duplicates

np.int64(0)

### Data Inspection Summary

The dataset contains **507,071 lightning events** recorded around **five airports** and described by **13 variables** including temporal (`date`), spatial (`lon`, `lat`), physical lightning characteristics (`amplitude`, `maxis`, `icloud`), and airport-related information (`dist`, `azimuth`, `airport`).  
Each row represents a single lightning strike detected within a **30 km radius around an airport**.  
The variables have different data types, including numerical variables (`float`, `int`), a categorical variable (`airport`), and a boolean variable (`icloud`).  
The `date` variable is currently stored as a string and will need to be converted to a **datetime format** to enable temporal analysis.  
Similarly, `airport_alert_id` is stored as a float due to the presence of missing values but will later be converted to a **nullable integer type (`Int64`)** since it represents an identifier.  
Most variables contain **no missing values**, except for `airport_alert_id` and `is_last_lightning_cloud_ground`, which contain about **89% missing values**.  
These missing values are expected because, according to the documentation, these variables are only recorded for lightning events occurring within **20 km of an airport**.  
Lightning events are unevenly distributed across the five airports, with **Pise, Bastia, and Biarritz** recording the highest number of observations.  
The amplitude values show a wide range of **positive and negative intensities**, corresponding to different lightning polarities, while the localization error (`maxis`) remains relatively small, indicating precise lightning detection.  
Finally, the variable `is_last_lightning_cloud_ground` identifies whether a lightning strike corresponds to the **last cloud-to-ground lightning of an alert sequence**, making it particularly relevant for the project objective of **predicting the end of lightning events**.

### Data Cleaning 

#### Handling missing values

The variables `airport_alert_id` and `is_last_lightning_cloud_ground` contain many missing values because they are only recorded for lightning events occurring within 20 km of an airport.  
For modeling purposes, `airport_alert_id` is converted to a nullable integer type (`Int64`), while missing values in `is_last_lightning_cloud_ground` are replaced with `False`, since lightning events outside the alert zone cannot correspond to the last lightning of an alert.

In [65]:
df["airport_alert_id"] = df["airport_alert_id"].astype("Int64")
df["is_last_lightning_cloud_ground"] = df["is_last_lightning_cloud_ground"].fillna(False)
df["is_last_lightning_cloud_ground"] = df["is_last_lightning_cloud_ground"].astype("bool")